implementations/rust/ockam/ockam_identity/src/channel/trust_options.rs

```diff
@@ -36,11 +54,55 @@ impl SecureChannelTrustOptions {
         self.trust_policy = Arc::new(trust_policy);
         self
     }
+
+    pub(crate) fn process(self, addresses: &Addresses) -> SecureChannelTrustOptionsProcessed {
```

Comment:
I think we are cramming too many things in there:

 - the trust policy is copied in the `SecureChannelTrustOptionsProcessed` but could also be passed as an independent parameter to `Decryptor::create_initiator`. Then we can call `SecureChannelTrustOptionsProcessed` `SecureChannelAccessControls` to mirror my suggestion of `TcpConnectionAccessControls` instead of `TcpConnectionTrustOptionsProcessed`

Then:

 - we are having 2 side-effects here
   - creating a session id
   - modifying the global `Sessions` state

I think it would be better to separate that in 2 functions:

 - `create_secure_channel_session(&self, sessions: &Sessions, addresses: &Addresses) -> Option<SessionId>`
    to update the sessions and return the session id
 -  `create_access_controls(&self, session_id) -> SecureChannelAccessControls`

I know that this entails some of the same checks (is there a `producer_session` etc...) but the data flow of what is modified and what is created would be clearer IMO.

Reply:
Yes those 2 functions can not be used separately. But that function looks innocuous and is actually modifying _shared global state_ in a non-trivial way so I would be more comfortable highlighting this aspect with a dedicated function and associated documentation (call that scars of years of Java code :-)). It's more about future readers / debuggers that I am thinking.